tap_snowflake/client.py

```diff
@@ -3,67 +3,228 @@
+                self.get_starting_timestamp(context=context)
+                if self.is_timestamp_replication_key
+                else self.get_starting_replication_key_value(context=context)
+            )
+            if start_val:
+                query = query.filter(replication_key_col >= start_val)
+
+        for record in self.connector.connection.execute(query):
+            yield dict(record)
+
+
+class SnowflakeStream(SelectableSQLStream):
     """Stream class for Snowflake streams."""
 
     connector_class = SnowflakeConnector
 
-    def get_records(self, partition: Optional[dict]) -> Iterable[Dict[str, Any]]:
-        """Return a generator of row-type dictionary objects.
+    def get_batches(
+        self, batch_config: BatchConfig, context: dict | None = None
+    ) -> Iterable[tuple[BaseBatchFileEncoding, list[str]]]:
+        """Get batches of Records from Snowflake.
 
-        Developers may optionally add custom logic before calling the default
-        implementation inherited from the base class.
+        Currently this returns records batches unloaded via an internal user stage.
+        In future this can be updated to include new methods for unloading via external stages.
+        For more details on batch unloading data from Snowflake, see the Snowflake docs:
+        https://docs.snowflake.com/en/user-guide-data-unload.html
+        """
+        if context:
+            raise NotImplementedError(
+                f"Stream '{self.name}' does not support partitioning."
+            )
+        return self.get_batches_from_internal_user_stage(batch_config, context)
 
-        Args:
-            partition: If provided, will read specifically from this data slice.
+    @staticmethod
+    def _get_full_table_copy_statement(
+        sync_id: str, prefix: str, objects: str, table_name: str
+    ) -> Tuple[text, dict]:
+        """Get FULL_TABLE copy statement and key bindings."""
+        return (
+            text(
+                f"copy into '@~/tap-snowflake/{sync_id}/{prefix}' from "
+                + f"(select object_construct({', '.join(objects)}) from {table_name}) "
+                + "file_format = (type='JSON', compression='GZIP') overwrite = TRUE"
+            ),
+            {},
+        )
 
-        Yields:
-            One dict per record.
+    def _get_incremental_copy_statement(
+        self,
+        sync_id: str,
+        prefix: str,
+        objects: str,
+        table_name: str,
+        table,
+        replication_key_value,
+    ) -> Tuple[text, dict]:
+        """Get INCREMENTAL copy statement and key bindings."""
+        return (
+            text(
+                f"copy into '@~/tap-snowflake/{sync_id}/{prefix}' from "
+                + f"(select object_construct({', '.join(objects)}) from {table_name} where {self.replication_key} >= :replication_key_value order by {self.replication_key})"
+                + "file_format = (type='JSON', compression='GZIP') overwrite = TRUE"
```

Comment:
@kgpayne - You have this correct here, but I wanted to amend my prior suggestion of enforcing an ending value here. I've done some exploration here in this comment, but essentially what I realized is that we can't cap that value unless we want to do a final batch without the upper bound. So, upshot, this is perfect. Let's keep signposts out of the upper bound logic, as you have here. ✅ 

- https://github.com/meltano/sdk/discussions/1011#discussioncomment-3756476